source/slang/decl-defs.h

```diff
@@ -180,6 +186,9 @@ SIMPLE_SYNTAX_CLASS(ScopeDecl, ContainerDecl)
 // A function/initializer/subscript parameter (potentially mutable)
 SIMPLE_SYNTAX_CLASS(ParamDecl, VarDeclBase)
 
+// A parameter of a function declared in "modern" types (immutable unless explicitly `out` or `inout`)
```

Comment:
Why not just 'data' on a ParamDecl as opposed to another type?

Reply:
I've tried (to differing levels of success) to keep from storing flags and other bits of non-structural state on AST needs. The IR does a better job of it, and shows the benefits more (the highly regular structure of the IR allows many algorithms to be written in a way that doesn't care about the nominal "class hierarchy"). The AST isn't set up as nicely for that kind of thing, but that *is* where I would like it to go in the long run.

Encoding the information that differentiates different cases in the hierarchy also makes it easier to enforce invariants via the type system. E.g., if I have a subroutine that only expects to work on modern-style parameters, I can express that in its parameter signature, rather than having to take a plain `ParamDecl*` and then assert that it is one I want to handle. In this case that isn't a strong argument.

*If* you buy my argument against using general-purpose flags/data, then the main alternative to using the class hierarchy to encode this stuff is using decorations. I'd definitely be in favor of using decorations in cases where, say, an AST node can be either an A or a B, *and* it can be either an X or a Y, but neither of those is really more fundamental to the class of the AST node than the other. Using something like an `IsModernStyleDecoration` in this case would be fine, but it would be pretty heavy-weight for such a simple case.

(Yes, I realize that you might advocate for storing a simple set of bitflags on every AST node, and then having support for the "does this node have such-and-such decoration" test to use those bitflags as a compacted representation of certain decorations. I'm fine with pursuing that at some point, but I'd like to consider it *after* we've done a pass to systematize and regularize the AST representation in preparation for serialization.)